### PR TITLE
[do not merge] Fix tx cache

### PR DESCRIPTION
**Pull request by flyingsilverfin**
## What is the goal of this PR?

Refactor some low level handling of `Transaction`, `SessionImpl` and caching mechanisms.

## What are the changes implemented in this PR?

For the time being we don't rely anymore on our custom Cache to reuse `SchemaConcept`s as this leads to exceptions when using these concepts in different transactions, instead we cache Labels for schema concepts and we rely on Janusgraph graph cache.

GraknClient:
  - when closing a session obtained from `new GraknClient(uri).session()` the HTTP2 will be no longer closed, the user will instead have to close the client invoking `graknClient.close()` in order to properly close the HTTP2 connection to the Server.
 - the `Grakn.Session` produced by the GraknClient now has a `isOpen` boolean (so that we cannot send multiple close requests to the server by mistake)

grakn.properties:
 - enabled by default Janus graph cache

other:
 - `SessionStore` renamed to `SessionFactory`
 - updated tests
 - `GlobalCache` renamed to `KeyspaceCache` as it holds labels for a specific keyspace schema
-  internal refactoring in `SessionFactory` and `KeyspaceManager` to better handle concurrency and maintainability

### CODE REVIEW

File: server/src/server/keyspace/KeyspaceManager.java [hunk 1]
```diff
@@ -57,7 +62,10 @@
 
     public KeyspaceManager(Config config){
         this.config = config;
-        this.systemKeyspaceSession = new SessionImpl(SYSTEM_KB_KEYSPACE, config);
+
```
Comment (flyingsilverfin) [inline server/src/server/keyspace/KeyspaceManager.java]: Creating new keyspace cache and JanusGraph here could be cleaner

File: server/src/server/rpc/SessionService.java [hunk 2]
```diff
@@ -268,9 +269,24 @@ private void open(Transaction.Open.Req request) {
                 throw ResponseBuilder.exception(Status.FAILED_PRECONDITION);
             }
 
-            tx = openSessions.get(request.getSessionId()).transaction(Type.of(request.getType().getNumber()));
+            ScopedSpan span = null;
```
Comment (flyingsilverfin) [inline server/src/server/rpc/SessionService.java]: Remove tracing in final PR merge

File: server/src/server/session/SessionFactory.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+        // problem option 1: map may have a new, valid entry inserted right after we null the value atomically but before we delete it
+        // so we end up deleting a valid entry
+
+        // problem option 2: remove entry from map, second thread creates a new entry and tries to open a new JanusGraph with
+        // the same name before we are able to delete the close/delete the old JanusGraph
+
+        // leave the nulled out value for now, further problems exist higher up the stack eg. possibly two concurrent hits on KeyspaceService.delete()?
+
+        KeyspaceCacheContainer keyspaceCacheContainer = keyspaceCacheMap.remove(keyspace);
```
Comment (flyingsilverfin) [inline server/src/server/session/SessionFactory.java]: Write a test to address concerns in comments if possible

File: server/src/server/session/SessionFactory.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+    /**
+     * Obtain a new session to a keyspace, while doing the required accounting
+     * for reference counting keyspace cache usage to ensure proper keyspaceCache eviction
+     *
+     * @param keyspace
+     * @param config
+     * @return
+     */
+    private SessionImpl newSessionImpl(Keyspace keyspace, Config config) {
```
Comment (flyingsilverfin) [inline server/src/server/session/SessionFactory.java]: Clean this up better

File: server/src/server/session/SessionImpl.java [hunk 5]
```diff
@@ -54,50 +61,113 @@
...
     /**
      * Instantiates {@link SessionImpl} specific for internal use (within Grakn Server),
      * using provided Grakn configuration
      *
      * @param keyspace to which keyspace the session should be bound to
-     * @param config   config to be used. If null is supplied, it will be created
+     * @param config   config to be used.
      */
-    public SessionImpl(Keyspace keyspace, Config config) {
+    public SessionImpl(Keyspace keyspace, Config config, KeyspaceCache keyspaceCache, JanusGraph graph, Runnable onClose) {
```
Comment (flyingsilverfin) [inline server/src/server/session/SessionImpl.java]: Review Sessionimpl in full, coulld be cleaner

File: server/src/server/session/SessionImpl.java [hunk 6]
```diff
@@ -54,50 +61,113 @@
...
+        }
+        copyMetaConceptsToKeyspaceCache(tx);
+        tx.commit();
+
+    }
 
     @Override
     public TransactionOLTP transaction(Transaction.Type type) {
+
+        ScopedSpan span = null;
```
Comment (flyingsilverfin) [inline server/src/server/session/SessionImpl.java]: remove tracing

File: server/src/server/session/TransactionOLTP.java [hunk 7]
```diff
@@ -101,45 +103,63 @@
...
     @Nullable
     private GraphTraversalSource graphTraversalSource = null;
 
-    public TransactionOLTP(SessionImpl session, JanusGraph janusGraph) {
+    public TransactionOLTP(SessionImpl session, JanusGraph janusGraph, KeyspaceCache keyspaceCache) {
+
+        createdInCurrentThread.set(true);
+
+        ScopedSpan span = null;
+        if (ServerTracingInstrumentation.tracingActive()) span = ServerTracingInstrumentation.createScopedChildSpan("TransactionOLTP constructor");
```
Comment (flyingsilverfin) [inline server/src/server/session/TransactionOLTP.java]: remove tracing

File: server/src/server/session/TransactionOLTP.java [hunk 8]
```diff
@@ -457,7 +412,7 @@ public void checkMutationAllowed() {
      * @param baseType The base type of the new type
      * @return The new type vertex
      */
-    private VertexElement addTypeVertex(LabelId id, Label label, Schema.BaseType baseType) {
+    protected VertexElement addTypeVertex(LabelId id, Label label, Schema.BaseType baseType) {
```
Comment (flyingsilverfin) [inline server/src/server/session/TransactionOLTP.java]: Revert to private

File: server/src/server/session/TransactionOLTP.java [hunk 9]
```diff
@@ -553,11 +512,11 @@ private TransactionException labelTaken(SchemaConcept schemaConcept) {
      * @return The {@link SchemaConcept} which was either cached or built via a DB read or write
      */
     private SchemaConcept buildSchemaConcept(Label label, Supplier<SchemaConcept> dbBuilder) {
-        if (cache().isTypeCached(label)) {
-            return cache().getCachedSchemaConcept(label);
-        } else {
+//        if (cache().isTypeCached(label)) {
```
Comment (flyingsilverfin) [inline server/src/server/session/TransactionOLTP.java]: Decide whether to keep commented as reminder of how to re-enable Concept Type cache if we need it, or whether to delete this code.

File: server/src/server/session/TransactionOLTP.java [hunk 10]
```diff
@@ -745,9 +701,10 @@ public void commit() throws InvalidKBException {
         }
         try {
             validateGraph();
-            commitTransactionInternal();
-            cache().writeToGraphCache(true);
-            //TODO update cache here
+            synchronized (keyspaceCache) {
```
Comment (flyingsilverfin) [inline server/src/server/session/TransactionOLTP.java]: Comment why this needs `synchronized`

File: server/src/server/session/TransactionOLTP.java [hunk 11]
```diff
@@ -790,9 +749,10 @@ private void closeTransaction(String closedReason) {
         Map<String, Set<ConceptId>> newAttributes = cache().getNewAttributes();
         boolean logsExist = !newInstances.isEmpty() || !newAttributes.isEmpty();
 
-        commitTransactionInternal();
-
-        cache().writeToGraphCache(true);
+        synchronized (keyspaceCache) {
```
Comment (flyingsilverfin) [inline server/src/server/session/TransactionOLTP.java]: Comment why `synchronized`

File: server/src/server/session/cache/KeyspaceCache.java [hunk 12]
```diff
@@ -68,6 +70,26 @@ public GlobalCache(Config config) {
                 .build();
     }
 
+    public KeyspaceCache(Map<Label, LabelId> cachedLabels, Cache<Label, SchemaConcept> cachedTypes) {
+        this.cachedLabels = cachedLabels;
+        this.cachedTypes =  cachedTypes;
+    }
+
+    public KeyspaceCache copy(int cacheTimeout) {
```
Comment (flyingsilverfin) [inline server/src/server/session/cache/KeyspaceCache.java]: Remove if not used

File: server/src/server/session/cache/KeyspaceCache.java [hunk 13]
```diff
@@ -127,7 +149,7 @@ void readTxCache(TransactionCache transactionCache) {
 
                 //Add a new one
                 cachedLabels.putAll(transactionCache.getLabelCache());
-                cachedTypes.putAll(transactionCache.getSchemaConceptCache());
+//                cachedTypes.putAll(transactionCache.getSchemaConceptCache());
```
Comment (flyingsilverfin) [inline server/src/server/session/cache/KeyspaceCache.java]: [cont] decide what to do with Concept Type cache

File: test-integration/rule/GraknTestServer.java [hunk 14]
```diff
@@ -200,14 +195,14 @@ private Server createServer() {
         keyspaceStore = new KeyspaceManager(serverConfig);
 
         // tx-factory
```
Comment (flyingsilverfin) [inline test-integration/rule/GraknTestServer.java]: Change or remove comment

File: test-integration/server/session/SessionCacheIT.java [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+            Set<String> entityTypeSubs = tx.getMetaEntityType().subs().map(et -> et.label().getValue()).collect(toSet());
+            assertTrue(entityTypeSubs.contains("animal"));
+            Set<String> relationshipTypeSubs = tx.getMetaRelationType().subs().map(et -> et.label().getValue()).collect(toSet());
+            assertTrue(relationshipTypeSubs.contains("test-relationship"));
+        }
+        testSession.close();
+    }
+
+    @Test
+    public void addEntityWithLocalSession_possibleToRetrieveItWithNewLocalSessionClosingPreviousOne(){
```
Comment (flyingsilverfin) [inline test-integration/server/session/SessionCacheIT.java]: Add comments to following tests

File: server/src/server/session/TransactionOLTP.java (hunk 11 above)
Comment (flyingsilverfin) [reply server/src/server/session/TransactionOLTP.java]: completed

**Comment by marco-scoppetta** [comment]
Closing this given that's conflicting, just resolved and cleaned conflicts on a PR from my fork: https://github.com/graknlabs/grakn/pull/4970

**Pull request closed without merging**